platform/cog-platform-fdo.c

```diff
@@ -1306,8 +1554,13 @@ clear_wayland (void)
     if (wl_data.shell != NULL)
         wl_shell_destroy (wl_data.shell);
 
-    if (wl_data.compositor != NULL)
-        wl_compositor_destroy (wl_data.compositor);
+    g_clear_pointer (&wl_data.subcompositor, wl_subcompositor_destroy);
+    g_clear_pointer (&wl_data.compositor, wl_compositor_destroy);
+
+#if COG_ENABLE_WESTON_DIRECT_DISPLAY
+    g_clear_pointer (&wl_data.protection, weston_content_protection_destroy);
+    g_clear_pointer (&wl_data.direct_display, weston_direct_display_v1_destroy);
+#endif
```

Comment:
If the modifiers data will be allocated from dynamic memory, it should be freed here.